### PR TITLE
[edn] Add transition from boot to SW mode

### LINKED ISSUE #19037: [edn] Don't propagate SW commands after a fatal error
bad029c9c7 fixes the propagation of SW commands, so that writes to the `SW_CMD_REQ` register don’t propagate in multiple Auto mode states. The default assignment in EDN’s main FSM still allows the propagation of SW commands, though, and these default assignments apply even to the fatal error state. **Thus, SW can send commands to CSRNG even if EDN is in the fatal error state.** A CSRNG command that returns `genbits` will get those bits stored in EDN’s internal FIFOs. EDN’s endpoints cannot access the entropy in the fatal error state, however, because the acknowledge FSM stops acknowledging requests. We don’t test this case currently, but we do have the `EdnFatalAlertNoRsp_A` assertion, which ensures that EDN does not acknowledge any requests in the fatal alert state.

**For D3, we should tweak the main FSM so that it only propagates SW commands when needed (and certainly not in the fatal alert state).**

### LINKED ISSUE #19655: [edn] Should SW commands be allowed in boot_req_mode
### Description

In the documentation it says :

> In boot-time request mode, the command sequence is fully hardware-controlled and no command customization is possible. In this mode, the EDN automatically issues a special reduced-latency instantiate command followed by the default generate commands. This means, for instance, that no personalization strings or additional data may be passed to the CSRNG application interface port in this mode. On exiting, the EDN issues an uninstantiate command to destroy the associated CSRNG instance.

However, in the RTL I couldn't find anything that keeps SW commands from being issued in boot_req_mode. I also observed SW commands in the tests while CTRL.BOOT_REQ_MODE was set to true. 

Looking at the RTL code I think it is possible that sw commands can be issued in boot mode. It just relies on the SM not being in any of the following states: AutoSendGenCmd, AutoSendReseedCmd, AutoAckWait, AutoDispatch, AutoCaptGenCnt, AutoCaptReseedCnt, BootSendGenCmd or BootLoadIns. Other than in those states, a sw command can be issued if (reg2hw.sw_cmd_req.qe & sw_cmd_valid) is equal to 1’b1. [This](https://github.com/lowRISC/opentitan/pull/17083/commits/793f168b1e18a00c6f2ce63389668af64b7b0981) commit also aligns with the stated above.

If we decide to change the RTL, [this](https://github.com/lowRISC/opentitan/pull/17083/commits/793f168b1e18a00c6f2ce63389668af64b7b0981) commit could be extended to exclude more states. But in my opinion a cleaner solution could be to add a signal in the main_sm, that is high whenever the sm is in the AutoLoadIns or the SW_Port_Mode state. This way we could disallow sw commands unless the above signal is high.

If we decide to keep things as they are we may consider to adapt the documentation.

Another thing to consider is that the SM is in the BootDone state when the boot sequence is done. It stays there until the EDN is disabled or an error occurs. We could also discuss adding a transition from BootDone to Idle when the CTRL.BOOT_REQ_MODE flag is disabled. This might make the RTL align more with the quote from the documentation below, since atm disabling CTRL.BOOT_REQ_MODE doesn't cause the SM to leave the BootDone state. To my observation, there is also no uninstatiate command being issued when exiting boot mode.

> Once firmware initialization is complete, it is important to exit this mode if the endpoints ever need FIPS-approved random values. This is done by either clearing the EDN_ENABLE field or clearing the BOOT_REQ_MODE field in [CTRL](https://opentitan.org/book/hw/ip/edn/doc/registers.html#ctrl) to halt the boot-time request state machine. Firmware must then wait for successful the shutdown of the state machine by polling the REQ_MODE_SM_STS field of the [SUM_STS](https://opentitan.org/book/hw/ip/edn/doc/registers.html#sum_sts) register.

FYI: @vogelpi @andreaskurth

### PR DESCRIPTION
**Pull request by h-filali**
This commit adds a transition from boot to SW mode after the
boot sequence is done and the ctrl.boot_req_mode flag is disabled.

For further information have a look at the individual commit messages.

This PR resolves #19655 
This PR resolves #19037

### CODE REVIEW

**Comment by h-filali** [comment]
There's at least two things that need to be changed but I want to discuss first:
1. It's probably best to store the boot uninstantiate command somewhere in a parameter. Should I just add one to the edn_pkg.sv?
2. I need to change the values for the SM states, since we want to have the max hamming distance between the values (which is 3). I found a script but wasn't quite able to get it to work. Is [this](https://github.com/lowRISC/opentitan/blob/master/util/design/sparse-fsm-encode.py) script still up to date?

**Comment by h-filali** [comment]
> There's at least two things that need to be changed but I want to discuss first:
> 
>     1. It's probably best to store the boot uninstantiate command somewhere in a parameter. Should I just add one to the edn_pkg.sv?
> 
>     2. I need to change the values for the SM states, since we want to have the max hamming distance between the values (which is 3). I found a script but wasn't quite able to get it to work. Is [this](https://github.com/lowRISC/opentitan/blob/master/util/design/sparse-fsm-encode.py) script still up to date?

Both of the above points are now solved in this PR. The states now have the maximum hamming distances and the script still works. Furthermore, I added a param for the uninstantiate command.

File: hw/ip/edn/rtl/edn_main_sm.sv
Comment (vogelpi) [inline hw/ip/edn/rtl/edn_main_sm.sv:86]: This change is in the wrong commit. It's unrelated to adding the transition to SW mode.

File: hw/ip/edn/rtl/edn_main_sm.sv
Comment (vogelpi) [inline hw/ip/edn/rtl/edn_main_sm.sv:146]: Same here and below.

File: hw/ip/edn/rtl/edn_core.sv
Comment (vogelpi) [inline hw/ip/edn/rtl/edn_core.sv:481]: `edn_pkg::BOOT_UNINSTANTIATE` is not defined int his commit. The commit is not atomic.

File: hw/ip/edn/rtl/edn_main_sm.sv
Comment (vogelpi) [inline hw/ip/edn/rtl/edn_main_sm.sv:131]: This doesn't seem right. IIUC, `boot_req_mode_i` is high as long as the boot mode is configured in the  control register. This means the transition will only happen after SW disables boot mode and not once the boot sequence is done.

File: hw/ip/edn/doc/registers.md
Comment (vogelpi) [inline hw/ip/edn/doc/registers.md:451]: All changes in this file belong to the commit where you change the state encodings / add more states.

File: hw/ip/edn/doc/theory_of_operation.md
Comment (vogelpi) [inline hw/ip/edn/doc/theory_of_operation.md:30]: One sentence per line please.

Comment (vogelpi) [reply hw/ip/edn/doc/theory_of_operation.md:30]: Also below.

File: hw/ip/edn/doc/theory_of_operation.md
Comment (vogelpi) [inline hw/ip/edn/doc/theory_of_operation.md:78]: This change is unrelated to this PR. I think they are already part of another one.

File: hw/ip/edn/doc/theory_of_operation.md
Comment (vogelpi) [inline hw/ip/edn/doc/theory_of_operation.md:33]: I think we need to discuss this as soon as your back. From the description of the PR, related issue and existing documentation, I understood that boot mode should be exited automatically once the boot sequence ends. What the implementation and this part of the documentation suggest is that the boot mode is exited only when software explicitly requests that. What do we want? Let's discuss next week.

**Review by vogelpi — CHANGES_REQUESTED** [review]
Thanks for your PR @h-filali .

I see the following issues with this PR:
1. It contains RTL changes, i.e., we can't merge it right now.
2. It doesn't solve all the things discussed in the related issue.
3. It's not clear to me whether this PR changes the RTL in the way we have agreed previously or not.

Let's discuss this next week once your back.

**Comment by vogelpi** [comment]
Thanks @msfschaffner for pinging me. This PR slipped my radar.

File: hw/ip/edn/doc/registers.md
Comment (h-filali) [reply hw/ip/edn/doc/registers.md:451]: Thanks, you're right. I'll change this in my next push.

File: hw/ip/edn/rtl/edn_main_sm.sv
Comment (h-filali) [reply hw/ip/edn/rtl/edn_main_sm.sv:86]: You're right, I'll add a separate commit in my next push.

File: hw/ip/edn/rtl/edn_main_sm.sv
Comment (h-filali) [reply hw/ip/edn/rtl/edn_main_sm.sv:131]: Let's discuss this in person.

File: hw/ip/edn/doc/theory_of_operation.md
Comment (h-filali) [reply hw/ip/edn/doc/theory_of_operation.md:77]: You're right, this doesn't belong here. I'll change it in the next push.

File: hw/ip/edn/rtl/edn_main_sm.sv
Comment (vogelpi) [reply hw/ip/edn/rtl/edn_main_sm.sv:86]: Perfect, thanks!

File: hw/ip/edn/rtl/edn_core.sv
Comment (vogelpi) [reply hw/ip/edn/rtl/edn_core.sv:481]: Perfect, thanks.

**Review by vogelpi — CHANGES_REQUESTED** [review]
LGTM, thanks @h-filali . I'll approve this once we can again merge RTL changes.

**Pull request merged**
